### PR TITLE
[POC] add es search to grpc gateway

### PR DESCRIPTION
**Pull request by nicpuppa**
## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes #

### CODE REVIEW

File: zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/GatewayGrpcService.java [hunk 1]
```diff
@@ -55,9 +56,12 @@
 
 public class GatewayGrpcService extends GatewayImplBase {
   private final EndpointManager endpointManager;
+  private final UserServices userServices;
 
-  public GatewayGrpcService(final EndpointManager endpointManager) {
+  public GatewayGrpcService(
+      final EndpointManager endpointManager, final UserServices userServices) {
     this.endpointManager = endpointManager;
+    this.userServices = userServices;
```
Comment (koevskinikola) [inline zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/GatewayGrpcService.java:64]: ❓ Why do we need the `UserService` instance here? From what I can see, it's only used in the `AuthenticationInterceptor` creation inside the `Gateway` class.

**Review by koevskinikola — COMMENTED** [review]
✅ @nicpuppa the POC looks good, I just left some notes and minor questions. Nice job 🚀

File: zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/GatewayGrpcService.java (hunk 1 above)
Comment (nicpuppa) [reply zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/GatewayGrpcService.java:64]: Just a leftover 👍

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/authorization/TestESGrpcGateway.java
[the remark this answers is not in the document]
Comment (nicpuppa) [reply zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/authorization/TestESGrpcGateway.java:1]: I'll add more tests coverage in the real PR

File: zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/Gateway.java [hunk 2]
```diff
@@ -93,20 +97,27 @@ public final class Gateway implements CloseableSilently {
...
   public Gateway(
       final GatewayCfg gatewayCfg,
       final SecurityConfiguration securityConfiguration,
       final BrokerClient brokerClient,
       final ActorSchedulingService actorSchedulingService,
-      final ClientStreamer<JobActivationProperties> jobStreamer) {
+      final ClientStreamer<JobActivationProperties> jobStreamer,
+      final SecurityContextProvider securityContextProvider,
+      final UserSearchClient userSearchClient,
+      final PasswordEncoder passwordEncoder) {
```
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/Gateway.java:110]: Even if it's just for the review only to remove it in a later commit. There's a lot of junk to sift through 😄

File: zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/AuthenticationInterceptor.java [hunk 3]
```diff
@@ -71,6 +99,21 @@ public <ReqT, RespT> Listener<ReqT> interceptCall(
     }
   }
 
+  private String getUsernameFromClaims(final Map<String, Object> claims) {
+    final var username = claims.get("preferred_username");
```
Comment (remcowesterhoud) [inline zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/AuthenticationInterceptor.java:103]: ❓ Where does this `preferred_username` come from?

File: zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/AuthenticationInterceptor.java [hunk 4]
```diff
@@ -55,6 +69,20 @@ public <ReqT, RespT> Listener<ReqT> interceptCall(
...
+            "Denying call {} as the user {} does not exist",
+            methodDescriptor.getFullMethodName(),
+            username);
+        return deny(
+            call,
+            Status.UNAUTHENTICATED
+                .augmentDescription("Expected a valid user, but user does not exist")
+                .withCause(new IllegalArgumentException("User does not exist")));
+      }
+      claims.put(AUTHORIZED_USER_KEY, user.get().userKey());
```
Comment (remcowesterhoud) [inline zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/AuthenticationInterceptor.java:85]: ❌ As discussed on Slack. When we receive a bearer token we don't have to query ES. We must just validate the token and set all the claims on the token that goes to Zeebe.

Querying ES is only relevant for basic auth.

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/it/util/AuthorizationsUtil.java [hunk 5]
```diff
@@ -117,6 +119,31 @@ public boolean shouldRetryRequest(final StatusCode statusCode) {
...
+                applier.put("Authorization", "Bearer %s".formatted(encoder.encode()));
+              }
+
+              @Override
+              public boolean shouldRetryRequest(final StatusCode statusCode) {
+                return false;
+              }
+            })
+        .build();
+  }
```
Comment (remcowesterhoud) [inline zeebe/qa/util/src/main/java/io/camunda/zeebe/it/util/AuthorizationsUtil.java:145]: ❌ You'll likely won't need this anymore

**Review by remcowesterhoud — COMMENTED** [review]
Great work, overall it looks good to me except for the parts we discussed on Slack. I tried to summarize it a bit in the comment here.

File: zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/AuthenticationInterceptor.java (hunk 3 above)
Comment (remcowesterhoud) [reply zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/AuthenticationInterceptor.java:103]: This'll have to go as we don't know what claims are on a user's token. See comment below this as well.

**Pull request closed without merging**